tensorflow_addons/seq2seq/sampler.py

```diff
@@ -201,12 +202,15 @@ def sample_ids_shape(self):
     def sample_ids_dtype(self):
         return tf.int32
 
-    def initialize(self, inputs, sequence_length=None):
+    def initialize(self, inputs, sequence_length=None, mask=None):
         """Initialize the TrainSampler.
 
         Args:
           inputs: A (structure of) input tensors.
-          sequence_length: An int32 vector tensor.
+          sequence_length: An int32 vector tensor. Note that the
+            sequence_length has higher priority than mask, so mask
+            will be ignored if both of them are provided.
```

Comment:
I feel we should throw error if both of them are provided, rather than silently ignore the mask.

Reply:
I agree with you. But as I mentioned earlier, the problem with throwing an error when both are provided is that Keras implicitly passes the `mask` attribute in the Layer's `.call(..)` method. 

So even if the user does not specify the mask attribute and wants to use the `sequence_length`-type invocation it could face our error considering both `sequence_length` and `mask`(automatically) are provided. Please consider the following example:
```python
# masked_input will have the ._keras_mask attribute internally, 
# thus the Keras framework will pass this attribute 
# as a mask argument in the upcoming layers.
masked_input = ...
decoder = BasicDecoder(.., TrainingSampler(), ...)
decoder(masked_input, initial_state=..., sequence_length=sequence_length)
```
*Note that the decoder calls the `TrainingSampler.initialize(...)` internally.*

As you can see, the user doesn't specify the mask argument, but we'll receive it the `TrainingSampler.initialize(...)` independent of the user intention. 
For the user to overcome this issue, s/he will need to remove the `_keras_mask` attr. from the input tensor. 
```python
# a working example
masked_input = ...
delattr(masked_input, '_keras_mask')
decoder = BasicDecoder(.., TrainingSampler(), ...)
decoder(masked_input, initial_state=..., sequence_length=sequence_length)
```

So considering this annoyance for the end user, should we still throw an error if both of them are provided?

Reply:
If I remember it correctly, the mask value is only provided by keras if user either have a masking layer in the model, or use embedding layer with mask_zero = True. If you don't have those specified, I think keras framework will just inject none as mask when invoking the call().

Reply:
Yes, I think that is the case too. However, the decoder's input is commonly produced by feeding the word ids through an embedding layer (with `mask_zero=True` attribute). Thus the user will probably face the problem mentioned above.

Despite that, I think there will be some cases that Keras unintentionally passed a not-`None` mask attribute to our function.

Reply:
If the inputs are from embedding layer with mask_zero=True, then the mask should carry the same information as sequence_length, and user don't need to specify seqence_length at all.

I don't think keras will unintentionally pass a no-none mask to layer. Do you have any example?

Reply:
Yes, in that user doesn't need to pass the  sequence_length (which the most frequent case). 
Actually, I was worried about some use case that the user wants to implement a customized behavior in which the seqeucne_length and mask aren't the same (i don't have an example in my mind though). But it'd definitely a rare case. So I will change the `initialize` method to throw an error if both are provided.